File: web/src/api/stream.ts
import { w3cwebsocket } from 'websocket';

import { http } from '@/lib/http.ts';
import { getBaseUrl } from '@/lib/service.ts';

// set video stream mode
export function setMode(mode: string) {
  const data = { mode };
  return http.post('/api/stream/mode', data);
}

// set stream quality
export function setQuality(quality: number) {
  const data = { quality };
  return http.post('/api/stream/quality', data);
}

// set stream GOP
export function setGop(gop: number) {
  const data = { gop };
  return http.post('/api/stream/gop', data);
}

// WebRTC H.264 connection
export function webrtcH264() {
  return getWebSocket('/api/stream/h264/webrtc');
}

// direct H.264 connection
export function directH264() {
  return getWebSocket('/api/stream/h264/direct');
}

function getWebSocket(url: string) {
  const baseUrl = getBaseUrl('ws');

  return new w3cwebsocket(`${baseUrl}${url}`);
}

File: server/router/stream.go
package router

import (
	"NanoKVM-Server/middleware"
	"NanoKVM-Server/service/stream"
	"NanoKVM-Server/service/stream/direct"
	"NanoKVM-Server/service/stream/mjpeg"
	"NanoKVM-Server/service/stream/webrtc"

	"github.com/gin-gonic/gin"
)

func streamRouter(r *gin.Engine) {
	service := stream.NewService()

	api := r.Group("/api").Use(middleware.CheckToken())

	api.POST("/stream/rate-control", service.SetRateControlMode) // set rate control mode
	api.POST("/stream/mode", service.SetMode)                    // set stream mode
	api.POST("/stream/quality", service.SetQuality)              // set stream quality / bit-rate
	api.POST("/stream/gop", service.SetGop)                      // set stream gop
	api.POST("/stream/fps", service.SetFps)                      // set stream fps

	api.GET("/stream/mjpeg", mjpeg.Connect)        // mjpeg stream
	api.GET("/stream/h264/direct", direct.Connect) // h264 stream (direct)
	api.GET("/stream/h264/webrtc", webrtc.Connect) // h264 stream (webrtc)
}
